File: backend/src/main/java/com/foodadvisor/service/CompetitorComparisonService.java
package com.foodadvisor.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.foodadvisor.dto.competitor.CompetitorComparisonRequest;
import com.foodadvisor.dto.competitor.CompetitorComparisonResponse;
import com.foodadvisor.dto.competitor.CompetitorMerchantVO;
import com.foodadvisor.entity.Merchant;
import com.foodadvisor.exception.ApiException;
import com.foodadvisor.mapper.CompetitorComparisonMapper;
import com.foodadvisor.mapper.MerchantMapper;
import com.foodadvisor.trace.AiTraceContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 周边竞品对比服务（EPIC-02 Story 6）
 *
 * 负责：
 * 1. 查询本店与竞品的统计数据（评分、好评率、差评问题等）
 * 2. 校验竞品的区域和品类合法性
 * 3. 调用 AI 服务生成对比分析文字
 * 4. 组装数据库统计 + AI 分析为前端可用的响应
 *
 * 验收准则对齐：
 * - AC-1: 用户从相近区域和相同/相似品类中选择 2~3 家竞品
 * - AC-2: 对比至少包含价格、评分、好评率和评价数量
 * - AC-3: AI 分析突出优势或短板，无明显差异时明确说明
 * - AC-4: 固定数据下对比数值与数据库一致
 * - AC-5: 不符合区域/品类限制的商家默认不可选，强制请求时拒绝
 * - AC-7: 商家只能发起与自己店铺相关的竞品对比
 */
@Service
public class CompetitorComparisonService {

    private static final Logger log =
            LoggerFactory.getLogger(CompetitorComparisonService.class);

    /** 候选竞品查询上限 */
    private static final int MAX_CANDIDATE_COMPETITORS = 20;

    private final CompetitorComparisonMapper comparisonMapper;
    private final MerchantMapper merchantMapper;
    private final AIClientService aiClientService;
    private final JdbcTemplate jdbcTemplate;
    private final ObjectMapper objectMapper;
    private final AiRequestTraceService traceService;

    public CompetitorComparisonService(
            CompetitorComparisonMapper comparisonMapper,
            MerchantMapper merchantMapper,
            AIClientService aiClientService,
            JdbcTemplate jdbcTemplate,
            ObjectMapper objectMapper
    ) {
        this(comparisonMapper, merchantMapper, aiClientService,
                jdbcTemplate, objectMapper, null);
    }

    @Autowired
    public CompetitorComparisonService(
            CompetitorComparisonMapper comparisonMapper,
            MerchantMapper merchantMapper,
            AIClientService aiClientService,
            JdbcTemplate jdbcTemplate,
            ObjectMapper objectMapper,
            AiRequestTraceService traceService
    ) {
        this.comparisonMapper = comparisonMapper;
        this.merchantMapper = merchantMapper;
        this.aiClientService = aiClientService;
        this.jdbcTemplate = jdbcTemplate;
        this.objectMapper = objectMapper;
        this.traceService = traceService;
    }

    // ============================================
    // 商家端 API
    // ============================================

    /**
     * 获取本店周边可选的候选竞品列表。
     *
     * 筛选条件：同区域 + 同品类（或同菜系），营业中的活跃商家，
     * 排除本店自身。按评分降序排列，最多返回 20 家。
     *
     * 前端可据此渲染竞品选择器（验收准则 AC-1/AC-5）。
     *
     * @param merchantId 本店 ID
     * @return 候选竞品的基础信息列表
     */
    public List<CompetitorMerchantVO> getCandidateCompetitors(Long merchantId) {
        // 1. 获取本店基础信息
        Merchant self = merchantMapper.selectById(merchantId);
        if (self == null) {
                                    throw new ApiException(HttpStatus.NOT_FOUND, "NOT_FOUND", "商家不存在");
        }

        // 2. 查询同区域同品类的候选竞品
        List<Map<String, Object>> rows = comparisonMapper.findNearbyCompetitors(
                merchantId,
                self.getRegionCode(),
                self.getCategory(),
                self.getCuisine(),
                MAX_CANDIDATE_COMPETITORS
        );

        // 3. 转换为 VO
        return rows.stream()
                .map(this::mapToBasicVO)
                .collect(Collectors.toList());
    }

    /**
     * 执行竞品对比分析（不含 AI 追踪上下文，向后兼容）。
     *
     * @see #performComparison(Long, CompetitorComparisonRequest, AiTraceContext)
     */
    public CompetitorComparisonResponse performComparison(
            Long merchantId,
            CompetitorComparisonRequest request
    ) {
        AiTraceContext context = traceService == null
                ? null
                : traceService.startTrace(null, null, null, "COMPETITOR_ANALYSIS");
        try {
            return performComparison(merchantId, request, context);
        } catch (RuntimeException exception) {
            if (context != null) {
                traceService.failTraceSafely(context,
                        "COMPETITOR_ANALYSIS_FAILED", exception.getMessage());
            }
            throw exception;
        }
    }

    /**
     * 执行竞品对比分析（完整流程）。
     *
     * 流程：
     * 1. 获取本店基础信息 + 验证合法性
     * 2. 校验竞品 ID 均在合法候选列表中（区域 + 品类限制）
     * 3. 查询所有参与对比商家的详细统计数据
     * 4. 调用 AI 服务生成对比分析
     * 5. 组装最终响应
     *
     * @param merchantId 本店 ID
     * @param request    前端请求（包含竞品 ID 列表）
     * @param context    AI 追踪上下文（可为 null）
     * @return 完整对比结果（统计数据 + AI 分析）
     */
    public CompetitorComparisonResponse performComparison(
            Long merchantId,
            CompetitorComparisonRequest request,
            AiTraceContext context
    ) {
        // ---- 1. 获取本店基础信息 ----
        Merchant self = merchantMapper.selectById(merchantId);
        if (self == null) {
                                    throw new ApiException(HttpStatus.NOT_FOUND, "NOT_FOUND", "商家不存在");
        }

        // ---- 2. 校验竞品合法性（验收准则 AC-5） ----
        List<Long> competitorIds = request.getCompetitorMerchantIds();
        if (competitorIds == null || competitorIds.isEmpty()) {
                        throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY, "INVALID_INPUT",
                    "请至少选择 1 家竞品商家");
        }
        if (competitorIds.size() > 3) {
            throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY, "INVALID_INPUT",
                    "最多选择 3 家竞品商家");
        }
        if (competitorIds.contains(merchantId)) {
            throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY, "INVALID_INPUT",
                    "不能选择本店作为竞品");
        }

        // 获取合法候选竞品列表
        List<Map<String, Object>> candidates = comparisonMapper.findNearbyCompetitors(
                merchantId,
                self.getRegionCode(),
                self.getCategory(),
                self.getCuisine(),
                MAX_CANDIDATE_COMPETITORS
        );
        Set<Long> validCompetitorIds = candidates.stream()
                .map(row -> ((Number) row.get("id")).longValue())
                .collect(Collectors.toSet());

        // 校验请求中的竞品 ID 都合法
        for (Long cid : competitorIds) {
            if (!validCompetitorIds.contains(cid)) {
                throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY, "INVALID_COMPETITOR",
                        "商家 ID=" + cid + " 不在可对比范围内"
                                + "（需要同区域、同品类且正常营业）");
            }
        }

        // ---- 3. 查询所有参与对比商家的详细统计数据 ----
        // 组装本店 + 竞品的 ID 列表
        List<Long> allMerchantIds = new ArrayList<>();
        allMerchantIds.add(merchantId);
        allMerchantIds.addAll(competitorIds);

        // 查询每家商家的完整数据
        List<CompetitorMerchantVO> merchantDataList = new ArrayList<>();
        // 本店排第一
        merchantDataList.add(buildMerchantData(merchantId, true));
        for (Long cid : competitorIds) {
            merchantDataList.add(buildMerchantData(cid, false));
        }

        // ---- 4. 调用 AI 服务生成对比分析 ----
        List<Map<String, Object>> aiRequestData = merchantDataList.stream()
                .map(this::toAiRequestMap)
                .collect(Collectors.toList());

        JsonNode aiResult;
        try {
            aiResult = context == null
                    ? aiClientService.generateCompetitorComparison(merchantId, aiRequestData)
                    : aiClientService.generateCompetitorComparison(merchantId, aiRequestData, context);
        } catch (Exception e) {
            log.error("竞品对比 AI 调用失败 merchantId={}: {}", merchantId, e.getMessage());
            // AI 调用失败时返回降级结果（只有统计数据，无 AI 分析）
            return CompetitorComparisonResponse.builder()
                    .merchantId(merchantId)
                    .comparisonStatus("FAILED")
                    .merchantData(merchantDataList)
                    .errorMessage("AI 分析服务调用失败：" + e.getMessage())
                    .build();
        }

        // ---- 5. 组装响应 ----
        List<CompetitorComparisonResponse.AiMerchantAnalysis> aiAnalyses =
                parseAiAnalyses(aiResult);

        return CompetitorComparisonResponse.builder()
                .merchantId(merchantId)
                .comparisonStatus(
                        aiResult.has("comparisonStatus")
                                ? aiResult.get("comparisonStatus").asText()
                                : "SUCCESS"
                )
                .merchantData(merchantDataList)
                .aiMerchantAnalyses(aiAnalyses)
                .aiSummaryText(
                        aiResult.has("summaryText") && !aiResult.get("summaryText").isNull()
                                ? aiResult.get("summaryText").asText()
                                : null
                )
                .aiImprovementSuggestions(parseStringList(aiResult, "improvementSuggestions"))
                .modelName(
                        aiResult.has("modelName") && !aiResult.get("modelName").isNull()
                                ? aiResult.get("modelName").asText()
                                : null
                )
                .businessTraceId(
                        aiResult.has("businessTraceId") && !aiResult.get("businessTraceId").isNull()
                                ? aiResult.get("businessTraceId").asText()
                                : null
                )
                .build();
    }

    // ============================================
    // 商家统计数据组装
    // ============================================

    /**
     * 查询单家商家的完整统计数据（供对比使用）。
     *
     * 查询内容包括：
     * - 基本信息（名称、类别、菜系、地址、人均、评分、评价数）
     * - 好评率
     * - 分项评分（口味、环境、服务）
     * - 高频正面标签
     * - 主要差评问题
     *
     * @param merchantId 商家 ID
     * @param isSelf     是否为本店
     * @return 商家对比数据 VO
     */
    private CompetitorMerchantVO buildMerchantData(Long merchantId, boolean isSelf) {
        // 基础统计
        Map<String, Object> basic = comparisonMapper.getMerchantBasicStats(merchantId);
        if (basic == null || basic.isEmpty()) {
            throw new ApiException(HttpStatus.NOT_FOUND, "NOT_FOUND",
                    "商家 ID=" + merchantId + " 不存在");
        }

        // 好评率
        BigDecimal positiveRate = comparisonMapper.getPositiveRate(merchantId);

        // 分项评分
        Map<String, Object> dimRatings = comparisonMapper.getDimensionRatings(merchantId);

        // 正面标签
        List<String> topPositiveTags = comparisonMapper.getTopPositiveTags(merchantId)
                .stream()
                .map(row -> String.valueOf(row.get("tag_name")))
                .collect(Collectors.toList());

        // 差评问题
        List<String> topNegativeIssues = comparisonMapper.getTopNegativeIssues(merchantId)
                .stream()
                .map(row -> String.valueOf(row.get("category_name")))
                .collect(Collectors.toList());

        return CompetitorMerchantVO.builder()
                .merchantId(merchantId)
                .merchantName(String.valueOf(basic.get("name")))
                .category(String.valueOf(basic.getOrDefault("category", "")))
                .cuisine(basic.get("cuisine") != null
                        ? String.valueOf(basic.get("cuisine")) : null)
                .address(basic.get("address") != null
                        ? String.valueOf(basic.get("address")) : null)
                .isSelf(isSelf)
                .averagePrice(toBigDecimal(basic.get("average_price")))
                .rating(toBigDecimal(basic.get("rating")))
                .reviewCount(basic.get("review_count") instanceof Number n
                        ? n.intValue() : 0)
                .positiveRate(positiveRate)
                .tasteRating(toBigDecimal(
                        dimRatings != null ? dimRatings.get("taste_avg") : null))
                .environmentRating(toBigDecimal(
                        dimRatings != null ? dimRatings.get("environment_avg") : null))
                .serviceRating(toBigDecimal(
                        dimRatings != null ? dimRatings.get("service_avg") : null))
                .topPositiveTags(topPositiveTags)
                .topNegativeIssues(topNegativeIssues)
                .build();
    }

    // ============================================
    // AI 结果解析
    // ============================================

    /**
     * 解析 AI 返回的 merchantAnalyses 数组。
     */
    private List<CompetitorComparisonResponse.AiMerchantAnalysis> parseAiAnalyses(
            JsonNode aiResult
    ) {
        JsonNode analysesNode = aiResult.get("merchantAnalyses");
        if (analysesNode == null || !analysesNode.isArray()) {
            return Collections.emptyList();
        }

        List<CompetitorComparisonResponse.AiMerchantAnalysis> result = new ArrayList<>();
        for (JsonNode item : analysesNode) {
            result.add(CompetitorComparisonResponse.AiMerchantAnalysis.builder()
                    .merchantId(item.has("merchantId")
                            ? item.get("merchantId").asLong() : null)
                    .merchantName(item.has("merchantName")
                            ? item.get("merchantName").asText() : null)
                    .strengths(parseStringList(item, "strengths"))
                    .weaknesses(parseStringList(item, "weaknesses"))
                    .overallAssessment(item.has("overallAssessment")
                            ? item.get("overallAssessment").asText() : "")
                    .build());
        }
        return result;
    }

    /**
     * 从 JsonNode 中解析字符串数组。
     */
    private List<String> parseStringList(JsonNode parent, String fieldName) {
        JsonNode arrayNode = parent.get(fieldName);
        if (arrayNode == null || !arrayNode.isArray()) {
            return Collections.emptyList();
        }
        List<String> result = new ArrayList<>();
        for (JsonNode item : arrayNode) {
            if (item.isTextual()) {
                result.add(item.asText());
            }
        }
        return result;
    }

    // ============================================
    // 工具方法
    // ============================================

    /**
     * 将 CompetitorMerchantVO 转换为发送给 AI 服务的 Map 格式。
     */
    private Map<String, Object> toAiRequestMap(CompetitorMerchantVO vo) {
        Map<String, Object> map = new java.util.LinkedHashMap<>();
        map.put("merchantId", vo.getMerchantId());
        map.put("merchantName", vo.getMerchantName());
        map.put("category", vo.getCategory());
        map.put("cuisine", vo.getCuisine());
        map.put("address", vo.getAddress());
        map.put("averagePrice", vo.getAveragePrice());
        map.put("rating", vo.getRating());
        map.put("reviewCount", vo.getReviewCount());
        map.put("positiveRate", vo.getPositiveRate());
        map.put("tasteRating", vo.getTasteRating());
        map.put("environmentRating", vo.getEnvironmentRating());
        map.put("serviceRating", vo.getServiceRating());
        map.put("topPositiveTags", vo.getTopPositiveTags() != null
                ? vo.getTopPositiveTags() : Collections.emptyList());
        map.put("topNegativeIssues", vo.getTopNegativeIssues() != null
                ? vo.getTopNegativeIssues() : Collections.emptyList());
        return map;
    }

    /**
     * 将竞品候选列表的 Map 行转换为基础 VO（仅基本信息，不包含详细统计）。
     */
    private CompetitorMerchantVO mapToBasicVO(Map<String, Object> row) {
        return CompetitorMerchantVO.builder()
                .merchantId(((Number) row.get("id")).longValue())
                .merchantName(String.valueOf(row.get("name")))
                .category(String.valueOf(row.getOrDefault("category", "")))
                .cuisine(row.get("cuisine") != null
                        ? String.valueOf(row.get("cuisine")) : null)
                .address(row.get("address") != null
                        ? String.valueOf(row.get("address")) : null)
                .averagePrice(toBigDecimal(row.get("average_price")))
                .rating(toBigDecimal(row.get("rating")))
                .reviewCount(row.get("review_count") instanceof Number n
                        ? n.intValue() : 0)
                .isSelf(false)
                .build();
    }

    /**
     * 将 Object 安全转为 BigDecimal。
     */
    private BigDecimal toBigDecimal(Object value) {
        if (value == null) return null;
        if (value instanceof BigDecimal bd) return bd;
        try {
            return new BigDecimal(value.toString())
                    .setScale(1, RoundingMode.HALF_UP);
        } catch (NumberFormatException e) {
            return null;
        }
    }
}

File: ai-service/app/services/competitor_comparison_service.py
"""
周边竞品对比服务（EPIC-02 Story 6）

根据 Spring Boot 传入的本店与竞品统计数据，调用大模型生成：
- 每家商家的优势/短板分析（相对于竞品群体）
- 横向对比总结（本店在竞品中的定位）
- 基于差距的改进建议

核心原则：
1. 所有分析必须基于传入的真实统计数据，禁止编造不存在的事实
2. 差距不明显时明确说明"无明显差异"，不做强行解读
3. 改进建议必须具体、可执行，针对统计中暴露的具体短板
"""
import logging
import uuid
from typing import List, Optional

from app.models.schemas import (
    CompetitorComparisonRequest,
    CompetitorComparisonResponse,
    CompetitorMerchantData,
    CompetitorSingleComparisonResult,
)
from app.services.llm_service import llm_service

logger = logging.getLogger(__name__)

# ============================================
# 竞品对比分析提示词
# ============================================
COMPETITOR_COMPARISON_PROMPT = """你是一位餐饮行业的数据分析顾问，专精于为商家提供竞品对比分析和经营建议。

## 输入数据说明
你会收到一家"本店"和1~3家"竞品"的结构化统计数据。字段含义：
- averagePrice: 人均消费（元）
- rating: 综合评分（0~5分）
- reviewCount: 评价总数
- positiveRate: 好评率（0~1，即正面评价占比）
- tasteRating: 口味评分均值（0~5分）
- environmentRating: 环境评分均值（0~5分）
- serviceRating: 服务评分均值（0~5分）
- topPositiveTags: 顾客高频提及的正面标签
- topNegativeIssues: 顾客主要投诉的问题

## 分析任务
请从以下角度对本店和竞品进行对比分析，返回严格 JSON：

### 1. 每家商家的分析（merchantAnalyses）
对每一家商家（包括本店），分析其中相对于其他几家的：
- **strengths（优势）**：该商家在哪些维度明显领先于其他商家（至少领先10%才算优势）
- **weaknesses（短板）**：该商家在哪些维度明显落后于其他商家（至少落后10%才算短板）
- **overallAssessment（综合评价）**：1~2 句话总结该商家在竞品中的定位

### 2. 横向对比总结（summaryText）
2~3 句话概括本店在这组竞品中的整体表现和定位。
如果本店在多数维度领先 → 给出领先结论
如果本店在多数维度落后 → 给出追赶建议
如果各有优劣     → 点明关键优势项和关键短板项
如果差距不明显   → 明确说明"各商家在核心指标上差异不大"

### 3. 改进建议（improvementSuggestions）
只针对**本店**给出改进建议，必须：
- 每条建议对应一个具体的数据短板（如"服务评分3.2低于竞品平均4.1"）
- 建议具体可执行（如"建议在周末高峰期增加1-2名服务人员"而非"提升服务"）
- 最多3条建议，排序从最关键到次要
- 如果本店各项指标均领先于竞品 → 建议侧重如何保持优势和拓展新优势
- 如果差距不明显 → 可提出差异化竞争方向

## 规则（违反任何一条都是错误）
1. **禁止编造**：所有分析必须严格基于输入数据中的数字和标签，不得编造输入中没有的内容
2. **差异不明显的处理**：当某维度差距小于10%时，不要强行判定为"优势"或"短板"，可以说"与竞品持平"
3. **正向表达**：即使是指出短板，也要用建设性表述，不用贬低性语言
4. **具体引用**：提到任何结论时，尽量引用具体数值（如"本店评分4.3 > 竞品均值3.8"）
5. **不做横向拉踩**：不贬低任何商家，焦点放在客观数据和建设性建议上

## 输出格式（严格 JSON）
{
  "merchantAnalyses": [
    {
      "merchantId": 1,
      "merchantName": "本店名称",
      "strengths": ["优势1：具体数据对比", "优势2：具体数据对比"],
      "weaknesses": ["短板1：具体数据对比"],
      "overallAssessment": "综合评价文字"
    }
  ],
  "summaryText": "横向对比总结，2~3句话",
  "improvementSuggestions": [
    "建议1：针对具体问题 + 可执行措施",
    "建议2：针对具体问题 + 可执行措施"
  ]
}

## 注意事项
- strengths 和 weaknesses 可以为空数组（当没有明显优势/短板时）
- 不要为了填充数组而编造不存在的优劣
- 语言风格：专业、简洁、建设性（面向商家用户）
"""


class CompetitorComparisonService:
    """周边竞品对比服务 — 用大模型分析数据 + 生成对比见解"""

    def _generate_trace_id(self) -> str:
        """生成 AI 调用追踪ID"""
        return f"trace-{uuid.uuid4().hex[:16]}"

    async def compare(self, request: CompetitorComparisonRequest) -> CompetitorComparisonResponse:
        """
        对传入的商家统计数据进行竞品对比分析。

        调用大模型分析统计数据并生成：
        - 每家商家的优势/短板
        - 横向对比总结
        - 本店改进建议

        统计数据和对比维度都由 Spring Boot 负责计算和提供，
        AI 只负责生成自然语言分析，不修改任何数值。
        """
        trace_id = self._generate_trace_id()

        system_prompt = (
            request.systemPrompt
            if request.systemPrompt and request.systemPrompt.strip()
            else COMPETITOR_COMPARISON_PROMPT
        )
        prompt_version = (
            request.promptVersion.strip()
            if request.promptVersion and request.promptVersion.strip()
            else "competitor-comparison:v1"
        )

        # ---- 构造给 LLM 的用户消息 ----
        user_message = self._build_user_message(request)

        try:
            result = await llm_service.chat_json(
                system_prompt=system_prompt,
                user_message=user_message,
                temperature=0.3,  # 适中温度，允许一定表达多样性但要忠实数据
                max_tokens=6000,
            )
        except Exception as e:
            logger.error(
                f"竞品对比分析失败 merchantId={request.merchantId}: {repr(e)}"
            )
            return CompetitorComparisonResponse(
                merchantId=request.merchantId,
                comparisonStatus="FAILED",
                modelName=f"fallback:{llm_service.model}",
                promptVersion=prompt_version,
                businessTraceId=trace_id,
                errorMessage=repr(e)[:500],
            )

        # ---- 解析并校验 LLM 返回结果 ----
        merchant_analyses = self._parse_merchant_analyses(
            result.get("merchantAnalyses", []), request.competitors
        )
        summary_text = str(result.get("summaryText", "")).strip() or None
        suggestions = self._parse_suggestions(result.get("improvementSuggestions", []))

        return CompetitorComparisonResponse(
            merchantId=request.merchantId,
            comparisonStatus="SUCCESS",
            merchantAnalyses=merchant_analyses,
            summaryText=summary_text,
            improvementSuggestions=suggestions,
            modelName=llm_service.model,
            promptVersion=prompt_version,
            businessTraceId=trace_id,
        )

    # ============================================
    # 构造 LLM 输入
    # ============================================

    def _build_user_message(self, request: CompetitorComparisonRequest) -> str:
        """
        将 CompetitorComparisonRequest 中的商家数据格式化为
        大模型可读的自然语言表格。
        """
        lines = [
            f"以下是一家名为\"本店\"的商家和其周边同类型竞品的统计数据。",
            f"本店是商家ID={request.merchantId}。",
            f"请对本店及竞品进行对比分析。\n",
        ]

        for i, c in enumerate(request.competitors):
            is_self = c.merchantId == request.merchantId
            label = "【本店】" if is_self else f"【竞品{i + 1}】"

            lines.append(f"{label} {c.merchantName}")
            lines.append(f"  - 商家ID: {c.merchantId}")
            lines.append(f"  - 类别: {c.category}")
            if c.cuisine:
                lines.append(f"  - 菜系: {c.cuisine}")
            if c.address:
                lines.append(f"  - 地址: {c.address}")
            if c.averagePrice is not None:
                lines.append(f"  - 人均消费: ¥{c.averagePrice}")
            if c.rating is not None:
                lines.append(f"  - 综合评分: {c.rating}/5")
            lines.append(f"  - 评价总数: {c.reviewCount}")
            if c.positiveRate is not None:
                lines.append(f"  - 好评率: {c.positiveRate:.0%}")
            if c.tasteRating is not None:
                lines.append(f"  - 口味评分: {c.tasteRating}/5")
            if c.environmentRating is not None:
                lines.append(f"  - 环境评分: {c.environmentRating}/5")
            if c.serviceRating is not None:
                lines.append(f"  - 服务评分: {c.serviceRating}/5")
            if c.topPositiveTags:
                lines.append(f"  - 高频正面标签: {'、'.join(c.topPositiveTags[:5])}")
            if c.topNegativeIssues:
                lines.append(f"  - 主要差评问题: {'、'.join(c.topNegativeIssues[:5])}")
            lines.append("")

        return "\n".join(lines)

    # ============================================
    # 解析 & 校验 LLM 输出
    # ============================================

    def _parse_merchant_analyses(
        self,
        raw: list,
        competitors: List[CompetitorMerchantData],
    ) -> List[CompetitorSingleComparisonResult]:
        """
        解析 LLM 返回的商家分析列表，校验 merchantId 的合法性，
        过滤掉不存在或编造的商家。
        """
        valid_ids = {c.merchantId for c in competitors}
        valid_names = {c.merchantId: c.merchantName for c in competitors}
        parsed = []
        seen_ids = set()

        for item in raw or []:
            if not isinstance(item, dict):
                continue

            mid = item.get("merchantId")
            if mid not in valid_ids or mid in seen_ids:
                continue  # 跳过非法或重复的商家ID

            seen_ids.add(mid)
            parsed.append(CompetitorSingleComparisonResult(
                merchantId=mid,
                merchantName=valid_names.get(mid, str(mid)),
                strengths=self._clean_string_list(item.get("strengths", [])),
                weaknesses=self._clean_string_list(item.get("weaknesses", [])),
                overallAssessment=str(item.get("overallAssessment", "")).strip(),
            ))

        # 确保所有传入的商家都有分析（LLM漏掉时补一条空记录）
        for c in competitors:
            if c.merchantId not in seen_ids:
                parsed.append(CompetitorSingleComparisonResult(
                    merchantId=c.merchantId,
                    merchantName=c.merchantName,
                    overallAssessment="暂无分析",
                ))

        return parsed

    def _parse_suggestions(self, raw: list) -> List[str]:
        """解析改进建议列表，过滤空字符串，最多保留3条"""
        suggestions = []
        for item in raw or []:
            if isinstance(item, str) and item.strip():
                suggestions.append(item.strip())
        return suggestions[:3]

    def _clean_string_list(self, raw: list) -> List[str]:
        """清洗字符串列表：去重、去空、保留顺序"""
        seen = set()
        result = []
        for item in raw or []:
            if isinstance(item, str) and item.strip():
                s = item.strip()
                if s not in seen:
                    seen.add(s)
                    result.append(s)
        return result


# 单例
competitor_comparison_service = CompetitorComparisonService()
